src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1404,18 +1383,43 @@ ucs_status_t uct_dc_mlx5_ep_check_fc(uct_dc_mlx5_iface_t *iface,
         return UCS_OK;
     }
 
-    status = uct_rc_fc_ctrl(&ep->super.super, UCT_RC_EP_FLAG_FC_HARD_REQ,
-                            NULL);
+    now = ucs_get_time();
+    it  = kh_put(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, (uint64_t)ep,
+                 &ret);
+    if (ucs_unlikely(ret == UCS_KH_PUT_FAILED)) {
+        ucs_error("failed to create hash entry for fc hard req");
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
+    } else if (ret == UCS_KH_PUT_KEY_PRESENT) {
+        fc_entry = &kh_value(&iface->tx.fc_hash, it);
+
+        /* Do not resend FC request if timeout is not reached, or window is
+         * still not empty */
+        if (ucs_likely((ep->fc.fc_wnd > 0) ||
+                       ((now - fc_entry->send_time) <
+                                iface->tx.fc_hard_req_timeout))) {
+            goto out_set_status;
+        }
+    } else {
+        ucs_assert(ep->fc.fc_wnd ==
+                           iface->super.super.config.fc_hard_thresh);
+        fc_entry = &kh_value(&iface->tx.fc_hash, it);
+    }
 
-    if ((ep->dci != UCT_DC_MLX5_EP_NO_DCI) && 
-        !uct_dc_mlx5_iface_is_dci_rand(iface)) {
-        ucs_assertv(uct_dc_mlx5_iface_dci_has_outstanding(iface, ep->dci),
-                    "iface %p ep %p: dci leak detected: dci=%d fc_wnd=%d",
-                    iface, ep, ep->dci, ep->fc.fc_wnd);
+    status = uct_dc_mlx5_ep_fc_hard_req_send(ep, fc_entry, now);
+    if (status != UCS_OK) {
+        if (ep->fc.fc_wnd > 0) {
+            kh_del(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, it);
+        }
+
+        goto out;
     }
 
-    return ((ep->fc.fc_wnd > 0) || (status != UCS_OK)) ?
-           status : UCS_ERR_NO_RESOURCE;
+out_set_status:
+    status = ucs_likely(ep->fc.fc_wnd > 0) ? UCS_OK : UCS_ERR_NO_RESOURCE;
+out:
+    uct_dc_mlx5_ep_check_dci_leak(ep);
```

Comment:
in this PR: instead of checking dci_leak, need to dci_put if we're going to return status!=UCS_OK - because if we return !=OK, it means called will not send anything, which may lead to leak

Reply:
update: don't dci_put since it will force re-allocating of dci while waiting for ep resources (flow control). just remove leak check from this PR.